### ISSUE TITLE
Error: datetime.timedelta is not JSON-serializable

### ISSUE THREAD

**Issue by wil93** [RedTurtle/eve-sqlalchemy#37]
Hi, I just started using eve-sqlalchemy but I get an error when trying to access an object which has a timedelta field like this:

    from sqlalchemy.types import Interval
    ...
    token_gen_interval = Column(
        Interval,
        CheckConstraint("token_gen_interval > '0 seconds'"),
        nullable=False,
        default=timedelta(minutes=30))
    ...

The error I get is:

    TypeError: datetime.timedelta(0, 1800) is not JSON serializable

**Comment by amleczko** [comment]
you need to provide a custom json encoder:

```python
from eve.io.base import BaseJSONEncoder
from eve_sqlalchemy import SQL as SQLAlchemy
from datetime import timedelta


class SQLAJSONEncoder(BaseJSONEncoder):
    def default(self, obj):
        if isinstance(obj, timedelta):
            return int(obj)
        return super(SQLAJSONEncoder, self).default(obj)


class CustomEve(SQLAlchemy):
    json_encoder_class = SQLAJSONEncoder


app = Eve(data=CustomEve)
```

**Comment by wil93** [comment]
Thank you, however I keep getting the same error... I modified the "simple example" from eve-sqlalchemy website in order to demonstrate the error: [this](https://gist.github.com/wil93/61f3418eb1c55ec2a882#file-without_custom_encoder-py) is the version that does not include the custom JSON encoder, and [this](https://gist.github.com/wil93/61f3418eb1c55ec2a882#file-with_custom_encoder-py) is the version that includes it. Both versions give the aforementioned TypeError.

**Comment by amleczko** [comment]
I will check it

**Comment by wil93** [comment]
Hi, any news about this? :smile:

I was thinking that this is probably related to nicolaiarocci/eve#267

**Comment by amleczko** [comment]
It shouldn't be related - eve-sqlalchemy depends on eve>=0.5

**Comment by kiniou** [comment]
There might be an issue with eve.utils.document_etag() function.
Debugging the execution of your script with pdb, I can see eve is trying to generate the etag by encoding attributes with a simple json.dumps() instead of json_encoder_class like in eve.render.render_json().

**Comment by kiniou** [comment]
:confused: After looking at imports in eve/utils.py, the dumps() function comes from `bson` and not from `json` so I'm not sure the custom json encoder class can be used here.

**Comment by wil93** [comment]
Now that you mention it, I put a `__import__("pdb").set_trace()` in the `default(self, obj)` function to see if it was called, and that point was never reached. Now this `bson` thing could explain that.

**Comment by kiniou** [comment]
@wil93 you should also change `int(obj)` by `obj.total_seconds()` since `int()` can't take a `timedelta` object directly.

**Comment by wil93** [comment]
Right, thanks :smile:

**Comment by kiniou** [comment]
This issue should be fixed in Eve 0.5.4 (cf. eve@72036276f2204d59e54a8a9fda5a6cc98f21143a).

**Comment by amleczko** [comment]
Confirmed - https://github.com/nicolaiarocci/eve/commit/72036276f2204d59e54a8a9fda5a6cc98f21143a fix the bug
